### ISSUE TITLE
OpenCart 3.0.3.8 PayPal Express Extension generates API Call Syntax Error

### ISSUE THREAD

**Issue by Luftschlosser** [opencart/opencart#12855]
Some time ago my PayPal Express Checkout stopped working in the Live Environment (seemingly out of the blue), while everything is still running smooth in Sandbox Mode. With help of the PayPal Support I was able to trace two Syntax Errors generated by the PayPal Express Extension in the NVP/SOAP API Call:

**1. Item Sizes:** (Confirmed by PayPal Support)

> This is what is being sent to PayPal in the payment request.
> paymentrequest_0_itemlengthvalue "80.00000000"
> paymentrequest_0_itemlengthunit "mm"
> paymentrequest_0_itemwidthvalue "55.00000000"
> paymentrequest_0_itemwidthunit "mm"
> paymentrequest_0_itemheightvalue "90.00000000"
> 
> [https://developer.paypal.com/api/nvp-so ... estmessage](https://developer.paypal.com/api/nvp-soap/set-express-checkout-nvp/#link-setexpresscheckoutrequestmessage)
> Per the documentation these values can only be a positive integer, which means instead of 90.000000000 it should be 90. If this information is being sent incorrectly it will cause buyers not to be able to pay.

**2. Empty Parameter "PAYMENTREQUEST_0_SHIPPINGAMT":** (Not confirmed yet)

> According to the Debug Log, the API Call also includes "PAYMENTREQUEST_0_SHIPPINGAMT":"" but with an empty string value. The real shipping costs are listed as "L_PAYMENTREQUEST_0_AMT1" later, but this empty Paramter also seems to cause an Issue with the PayPal API.

This is the Full API Call from the Debug Log:

```
2023-10-25 16:55:58 - PayPal Express debug (Call data):
{"METHOD":"SetExpressCheckout","MAXAMT":19.64,
"RETURNURL":"https:\/\/vf-kinetics.de\/index.php?route=extension\/payment\/pp_express\/checkoutReturn",
"CANCELURL":"https:\/\/vf-kinetics.de\/index.php?route=checkout\/checkout",
"REQCONFIRMSHIPPING":0
,"NOSHIPPING":0,
"LOCALECODE":"EN",
"LANDINGPAGE":"Login",
"HDRIMG":"https:\/\/vf-kinetics.de\/image\/cache\/catalog\/Icon-750x90.png",
"PAYFLOWCOLOR":"",
"CHANNELTYPE":"Merchant",
"ALLOWNOTE":"0",
"PAYMENTREQUEST_0_SHIPTONAME":"Name",
"PAYMENTREQUEST_0_SHIPTOSTREET":"Info",
"PAYMENTREQUEST_0_SHIPTOSTREET2":"",
"PAYMENTREQUEST_0_SHIPTOCITY":"City",
"PAYMENTREQUEST_0_SHIPTOSTATE":"State",
"PAYMENTREQUEST_0_SHIPTOZIP":"00000",
"PAYMENTREQUEST_0_SHIPTOCOUNTRYCODE":"DE",
"ADDROVERRIDE":1,
"PAYMENTREQUEST_0_SHIPPINGAMT":"",
"PAYMENTREQUEST_0_CURRENCYCODE":"EUR",
"PAYMENTREQUEST_0_PAYMENTACTION":"Sale",
"L_PAYMENTREQUEST_0_DESC0":"",
"L_PAYMENTREQUEST_0_NAME0":"Merchandise Bundle",
"L_PAYMENTREQUEST_0_NUMBER0":"501-02-10",
"L_PAYMENTREQUEST_0_AMT0":11,
"L_PAYMENTREQUEST_0_QTY0":"1",
"L_PAYMENTREQUEST_0_ITEMURL0":"https:\/\/vf-kinetics.de\/index.php?route=product\/product&product_id=84",
"L_PAYMENTREQUEST_0_ITEMLENGTHVALUE0":"80.00000000",
"L_PAYMENTREQUEST_0_ITEMLENGTHUNIT0":"mm",
"L_PAYMENTREQUEST_0_ITEMWIDTHVALUE0":"55.00000000",
"L_PAYMENTREQUEST_0_ITEMWIDTHUNIT0":"mm",
"L_PAYMENTREQUEST_0_ITEMHEIGHTVALUE0":"90.00000000",
"L_PAYMENTREQUEST_0_ITEMHEIGHTUNIT0":"mm",
"L_PAYMENTREQUEST_0_NUMBER1":"shipping",
"L_PAYMENTREQUEST_0_NAME1":"Germany  (Weight: 1.00g)",
"L_PAYMENTREQUEST_0_AMT1":0.85,
"L_PAYMENTREQUEST_0_QTY1":1,
"L_PAYMENTREQUEST_0_NUMBER2":"tax",
"L_PAYMENTREQUEST_0_NAME2":"USt (19%)",
"L_PAYMENTREQUEST_0_AMT2":2.25,
"L_PAYMENTREQUEST_0_QTY2":1,
"PAYMENTREQUEST_0_ITEMAMT":"14.10",
"PAYMENTREQUEST_0_AMT":"14.10",
"L_BILLINGTYPE0":"RecurringPayments",
"L_BILLINGAGREEMENTDESCRIPTION0":"14.1 EUR every 1 month"}

2023-10-25 16:55:59 - PayPal Express debug (Result): "TOKEN=EC%2d5F900598AP103820F&TIMESTAMP=2023%2d10%2d25T14%3a55%3a59Z&CORRELATIONID=a0d302a5a98da&ACK=Success&VERSION=109%2e0&BUILD=58205780"
```

There is no Error Reported by the API (ACK=Success), but the PayPal Window shows "An Error occured" (no more info) and the payment isn't processed.

**Comment by Khnaz35** [comment]
@Luftschlosser The PayPal integration has been updated, and a new release has been made available for OpenCart version 3.0.3.9.

**Comment by ADDCreative** [comment]
@Luftschlosser The PayPal Express payment extension hasn't been updated for may years and has been removed from the latest versions. You will find the parts you need to change are the following locations. If you need help, it would be best to post on https://forum.opencart.com/.

https://github.com/opencart/opencart/blob/8b4144edc3d54bfabea1295f936f7962e3a81d59/upload/catalog/model/extension/payment/pp_express.php#L72

https://github.com/opencart/opencart/blob/8b4144edc3d54bfabea1295f936f7962e3a81d59/upload/catalog/model/extension/payment/pp_express.php#L118-L123

**Comment by Luftschlosser** [comment]
@ADDCreative Thank you for pointing me into the right direction! I was able to correct the pp_express.php file. But there are more issues it seems. According to the PayPal support I'm in contact with, I also need to do something about special characters as well as Parameter LOCALECODE:

> It looks like there are special characters in the shiptoname and shiptostreet:
> 
> paymentrequest_0_shiptoname    "Peter F\u00f6rtsch"
> paymentrequest_0_shiptostreet    "M\u00d6chtestrasse!" 
> paymentrequest_0_name    "Postcards A6" "Germany (Weight: 1.00g)" "USt (19%)"
> 
> Can you try to update the payment request to remove all of the special characters such as (19%). Also remove the exclamation point in the street address. Remove the \ in the shiptoname. 
> 
> "LOCALECODE":"EN",
> 
> That isn't a valid locale code, here is a listing of valid locale codes:
> https://developer.paypal.com/reference/locale-codes/ 
> EN is not listed as a region if you are wanting English it should be en_US. 
> 
> Sandbox is more forgiving of special characters and Locale Codes, but Live if there are special characters it can cause issues with the payment request. Especially quotes, & ! \ '. 

However, I'm not able to find where the LOCALECODE is being set in PHP. Could you maybe also point out the correct file and line?

...At this point I'm not so sure if the PayPal Support is right, I think special characters are not uncommon in european names. What should I do about them? I'm not sure they are a real problem, but In order to get to the root of this issue I think I need to get rid of all syntax errors, and the PayPal API documentation is very clear in that regard.
Maybe it would make sense to update the pp_express.php file after all? I know it has been removed from the latest OpenCart version, but still everyone in the Forum seems to use it and it's one of the few payment options which support recurring payments.

**Comment by ADDCreative** [comment]
@Luftschlosser You will find LOCALECODE at the following locations. It does look to be set to an invalid value. It's optional, so could just remove it and let PayPal select the locale.

https://github.com/opencart/opencart/blob/8b4144edc3d54bfabea1295f936f7962e3a81d59/upload/catalog/controller/extension/payment/pp_express.php#L77

https://github.com/opencart/opencart/blob/8b4144edc3d54bfabea1295f936f7962e3a81d59/upload/catalog/controller/extension/payment/pp_express.php#L1316

According to the documentation PayPal expects UTF-8 characters. Which is what OpenCart is sending.
https://developer.paypal.com/api/nvp-soap/NVPAPIOverview/#link-utfcharacterencoding

I'm not sure where the \u escaping is coming from, it's probably not OpenCart. Could be PayPal added that in the message or it's being added to your data somehow. Looking at the order in your database will tell you.

**Comment by mhcwebdesign** [comment]
PayPal Express was removed in OC 3.0.x.x, please try the PayPal Checkout Integration V2.2.0 extension available in the latest 3.0.x.x branch.
